### PR TITLE
[19595] Typeobjects with tests

### PR DESCRIPTION
**Pull request by adriancampo**
Branch created for merging typeobjects and their tests.
Needs: https://github.com/eProsima/IDL-Parser/pull/103

### CODE REVIEW

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 1]
```diff
@@ -0,0 +1,1101 @@
...
+    NameHash $flag.name$_$relation_name$_flag_name_hash = $bitmask.name$_$relation_name$_detail.name_hash();
+    MD5 $flag.name$_$relation_name$_name_hash("$flag.name$");
+    for(int i = 0;i < 4;++i)
+    {
+        ASSERT_EQ($flag.name$_$relation_name$_name_hash.digest[i], $flag.name$_$relation_name$_flag_name_hash[i]);
+    }
+    }
+    $endif$
+
+    flagId++;
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:572]: The flag positions do not have to be sequential. For instance, the [test IDL](https://github.com/eProsima/dds-types-test/blob/main/IDL/helpers/basic_inner_types.idl#L17)

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 2]
```diff
@@ -0,0 +1,1101 @@
...
+    $relation_capname$BitfieldSeq $bitset.name$_$relation_name$_bitfieldseq = $bitset.name$_$relation_name$_bitset.field_seq();
+>>
+
+check_bitfield_type(ctx, bitfield, relation, relation_name, relation_capname) ::= <<
+    //=====check_bitfield_type - $relation_name$ - $bitfield.name$=====
+    $relation_capname$Bitfield field_$bitset.name$_$relation_name$_$bitfield.name$ = $bitset.name$_$relation_name$_bitfieldseq.at(bitfieldId);
+    CommonBitfield $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield = field_$bitset.name$_$relation_name$_$bitfield.name$.common();
+    
+    unsigned short $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield_position = $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield.position();
+    ASSERT_EQ(bitfieldId, $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield_position);
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:609]: This logic does not have sense. Bitfields are characterized by an initial position and by a bitcount, so the next position must be at least `position + bitcount`.

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 3]
```diff
@@ -0,0 +1,1101 @@
...
+    //BitsetMemberFlag flags;// Unused. No flags apply
+    octet $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield_bitcount = $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield.bitcount();
+    ASSERT_EQ($bitfield.spec.bitSize$, $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield_bitcount);
+    TypeKind $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield_holder_type = $bitset.name$_$relation_name$_common_$bitfield.name$_bitset_bitfield.holder_type();// Must be primitive integer type
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:613]: An additional check is required to confirm that the holder type is correct.

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 4]
```diff
@@ -0,0 +1,1101 @@
...
+    $elseif(parameter.typecode.isWStringType)$
+    ASSERT_EQ("$parameter.value$", $annotation.name$_$relation_name$_$parameter.name$_default_value.string16_value());
+    $endif$
+    memberId++;
+>>
+
+check_const_type(ctx, const, relation, relation_name, relation_capname) ::= <<
+    //=====check_const_type - $relation_name$ - $const.name$=====
+    $get_type_identifier(ctx=ctx, type=const, relation=relation, relation_name=relation_name, var_name=["type_identifier"])$
+    ASSERT_EQ($const.name$_$relation_name$_type_identifier->_d(), $const.name$_$relation_name$_type_identifier->_d());
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:729]: Additional checks can be perform to check those types that might be const.

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 5]
```diff
@@ -0,0 +1,1101 @@
...
+>>
+
+check_array_type(ctx, array, relation, relation_name, relation_capname) ::= <<
+    //=====check_array_type - $relation_name$ - $array.name$=====
+    $get_type_identifier(ctx=ctx, type=array, relation=relation, relation_name=relation_name, var_name=["type_identifier"])$
+    ASSERT_EQ(eprosima::fastrtps::types::EK_$if(relation)$COMPLETE$else$MINIMAL$endif$, $array.name$_$relation_name$_type_identifier->_d());
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:735]: This is not correct. Array TypeIdentifier has `TI_PLAIN_ARRAY_SMALL` or `TI_PLAIN_ARRAY_LARGE` as discriminators. Please, fix also with the other collection types (sequences and maps).

These TypeIdentifiers should be checked as well.

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 6]
```diff
@@ -0,0 +1,1101 @@
...
+    
+    if($array.name$_$relation_name$_totalsize <= 100){
+        SBoundSeq $array.name$_$relation_name$_s_bound_seq;
+        $member.typecode.dimensions:{ dim | $array.name$_$relation_name$_s_bound_seq.push_back($dim$);};separator="\n"$
+        ASSERT_EQ($array.name$_$relation_name$_s_bound_seq, $array.name$_$relation_name$_type_identifier.array_sdefn().array_bound_seq());
+    }else{
+        LBoundSeq $array.name$_$relation_name$_l_bound_seq;
+        $member.typecode.dimensions:{ dim | $array.name$_$relation_name$_l_bound_seq.push_back($dim$);};separator="\n"$
+        ASSERT_EQ($array.name$_$relation_name$_l_bound_seq, $array.name$_$relation_name$_type_identifier.array_ldefn().array_bound_seq());
+    }
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:791]: This is not correct. It does not depend on the sum of the elements in each dimension, but on the number of elements of a specific dimension being greater than 255.

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 7]
```diff
@@ -0,0 +1,1101 @@
...
+test_typedef(ctx, parent, typedefs) ::= <<
+TEST(TypeObjectTests, TestTypedefTypeObject_$typedefs.name$)
+{
+    //Minimal
+    $check_alias_type(ctx=ctx, typedefs=typedefs, relation=false, relation_name="minimal", relation_capname="Minimal")$
+    $check_alias_related_type(ctx=ctx, typedefs=typedefs, relation=false, relation_name="minimal", relation_capname="Minimal")$
+
+    //Complete
+    $check_alias_type(ctx=ctx, typedefs=typedefs, relation=true, relation_name="complete", relation_capname="Complete")$
+    $check_alias_related_type(ctx=ctx, typedefs=typedefs, relation=true, relation_name="complete", relation_capname="Complete")$
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:922]: With the new API, this distinction is not going to apply.

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 8]
```diff
@@ -0,0 +1,1101 @@
...
+TEST(TypeObjectTests, TestConstTypeObject_$const.name$)
+{
+    //Minimal
+    $check_const_type(ctx=ctx, const=const, relation=false, relation_name="minimal", relation_capname="Minimal")$ 
+
+    //Complete
+    $check_const_type(ctx=ctx, const=const, relation=true, relation_name="complete", relation_capname="Complete")$ 
+}
+>>
+
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:1050]: I am missing the collection tests: sequences, arrays and maps.

File: src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg [hunk 9]
```diff
@@ -0,0 +1,1101 @@
...
+
+member_type(ctx, member, type_member, declarators) ::= <<>>
+element_type(ctx, element, type_element, declarator) ::= <<>>
+sequence_type(ctx, sequence, type_sequence) ::= <<>>
+map_type(ctx, map, key_type, value_type) ::= <<>>
+string_type(ctx, string) ::= <<>>
+wide_string_type(ctx, wstring) ::= <<>>
+array_declarator(ctx, array, array_type) ::= <<>>
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/fastdds/idl/templates/TypeObjectTestingTestSource.stg:1101]: This is not following the custom stg template. Please, fix.

Also, the sequence, map and array should call the corresponding tests.

Finally, the tests checking the string/wstring TypeIdentifiers are not included.

**Comment by JLBuenoLopez-eProsima** [comment]
@adriancampo, you might find every error that String Template is complaining about by running the tests. Before failing the tests due to not having the TypeObject implementation, the test code generation complains and breaks.

**Comment by JLBuenoLopez-eProsima** [comment]
CI fails generating Fast DDS IDLs but after having built Fast DDS-Gen.

**Comment by adriancampo** [comment]
Generation and compilation phases working.
Must check tests results later.

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM with approval of @adriancampo

**Pull request merged**